pkg/alertmanager/alertmanager_metrics.go

```diff
@@ -94,7 +98,7 @@ func newAlertmanagerMetrics() *alertmanagerMetrics {
 		numFailedNotifications: prometheus.NewDesc(
 			"cortex_alertmanager_notifications_failed_total",
 			"The total number of failed notifications.",
-			[]string{"user", "integration"}, nil),
+			[]string{"user", "integration", "reason"}, nil),
```

Comment:
This metric can already be very high cardinality with many tenants, so I'm wondering whether we should actually intentionally aggregate away `reason`. However, it can only be one of three values, (`other`, `clientError` or `serverError`), we could potentially accept it, and because this metric is only exposed for non-zero values, the increase in series would normally be low.

I'm trying to think how we would use it:

For investigating failures:
The extra detail is probably too little; I would just look at logs.

For alerting on failures as an operator:
There could be value if we only wanted to alert on a particular class of error. However, I'm not sure I trust all integrations and backend services to correctly return 4xx vs 5xx (think: arbitrary webhooks). It appears the original [intention](https://github.com/prometheus/alertmanager/pull/2927#issuecomment-1241042716) of the change was to alert on 5xx class errors specifically from the SNS integration, so the people doing that presumably have sufficient trust in the error codes returned from it.

For alerting on failures as a tenant/user:
I care about all errors, regardless of 4xx vs 5xx.

Reply:
I've checked some of our internal environments, even with the high number of tenants, the number of series for this metric is very small, so I don't think adding `reason` will be a problem.

Reply:
There is much discussion here (including mirroring your point about cardinality): https://github.com/prometheus/alertmanager/pull/2927

Overall, the main benefit is for monitoring _specific_ notifiers. While I agree that many systems have messy 400/500 semantics, if a specific notifier was known to not do this, it could be useful for alerting. SNS was the specific example given in the thread.

Another potential use case is if there's an outage in a 3rd party system where we happen to have an integration. The metrics can give operators some extra confidence (that they might be able to pass on to users) of "our deployment is working fine, but <X> is down."